### PR TITLE
ARROW-12420: [C++/Dataset] Reading null columns as dictionary not longer possible

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: cpp/src/arrow/array/array_dict.cc [hunk 1]
```diff
@@ -80,7 +80,6 @@ int64_t DictionaryArray::GetValueIndex(int64_t i) const {
 DictionaryArray::DictionaryArray(const std::shared_ptr<ArrayData>& data)
     : dict_type_(checked_cast<const DictionaryType*>(data->type.get())) {
   ARROW_CHECK_EQ(data->type->id(), Type::DICTIONARY);
-  ARROW_CHECK_NE(data->dictionary, nullptr);
```
Comment (kszucs) [inline cpp/src/arrow/array/array_dict.cc:83]: `array->dictionary()` lazily initializes the dictionary if it is not set

Comment (kszucs) [inline cpp/src/arrow/array/array_dict.cc:83]: This must be restored.

**Comment by kszucs** [comment]
I reproduced the problem using a simple cast test from null array to a dictionary array:

```cpp
TEST(Cast, FromNullToDictionary) {
  auto from = std::make_shared<NullArray>(10);
  auto to_type = dictionary(int8(), boolean());

  ASSERT_OK_AND_ASSIGN(auto expected, MakeArrayOfNull(to_type, 10));
  CheckCast(from, expected);
}
```

Somewhere `ArrayData::dictionary` gets set to nullptr or not set at all. I don't have a solution yet, it's a but hard to debug due to the Datum variant.

**Review by wesm — APPROVED** [review]
+1 here, merge on green build

**Pull request closed without merging**